neuromation/api/storage.py

```diff
@@ -337,14 +342,22 @@ def _uri_to_path(self, uri: URL) -> str:
         except neuromation.api.IllegalArgumentError:
             raise NotADirectoryError(errno.ENOTDIR, "Not a directory", str(dst))
         progress.enter(StorageProgressEnterDir(src, dst))
-        folder = sorted(src_path.iterdir(), key=lambda item: (item.is_dir(), item.name))
+        tasks = []
+        async with self._file_sem:
+            folder = sorted(
+                src_path.iterdir(), key=lambda item: (item.is_dir(), item.name)
+            )
         for child in folder:
             name = child.name
             if child.is_file():
-                await self._upload_file(src_path / name, dst / name, progress=progress)
+                tasks.append(
+                    self._upload_file(src_path / name, dst / name, progress=progress)
```

Comment:
Small note: please create tasks explicitly, e.g.
```
                tasks.append(loop.create_task(
                    self._upload_file(src_path / name, dst / name, progress=progress)
                 ))
```
1. `tasks` variable contains task objects after the change, not coroutines.
2. In my experince, I've found that asyncio tricks for implicit wrapping coroutines in tasks by `ensure_future()` and `gather()` is confusing.

Reply:
Since I use a helper function to avoid a code duplication, I prefer to keep this list containing coroutines and convert them to tasks in the helper function. To avoid repetitions.

Reply:
Yes, I see. Good choice.